apps/prairielearn/src/lib/course.ts

```diff
@@ -0,0 +1,197 @@
+import { syncDiskToSqlWithLock } from '../sync/syncFromDisk';
+import { IdSchema } from './db-types';
+import {
+  getCommitHashAsync,
+  getOrUpdateCourseCommitHashAsync,
+  updateCourseCommitHashAsync,
+} from './courseUtil';
+
+const sql = sqldb.loadSqlEquiv(__filename);
+
+/**
+ * Check that an assessment_instance_id really belongs to the given course_instance_id
+ *
+ * @param assessment_instance_id - The assessment instance to check.
+ * @param course_instance_id - The course instance it should belong to.
+ */
+export async function checkBelongsAsync(
+  assessment_instance_id: string,
+  course_instance_id: string,
+): Promise<void> {
+  if (
+    sqldb.queryOptionalRow(
+      sql.check_belongs,
+      {
+        assessment_instance_id,
+        course_instance_id,
+      },
+      IdSchema,
+    ) == null
+  ) {
+    throw new Error('access denied');
+  }
+}
+export const checkBelongs = callbackify(checkBelongsAsync);
+
+/**
+ * Return the name and UID (email) for every owner of the specified course.
+ *
+ * @param {string} course_id The ID of the course.
+ * @returns {Promise<{ uid: string, name?: string }[]>}
+ */
+export async function getCourseOwners(
+  course_id: string,
+): Promise<{ uid: string; name?: string }[]> {
+  const { rows } = await sqldb.queryAsync(sql.select_owners, { course_id });
+  return rows.map((row) => ({
+    uid: row.uid,
+    name: row.name,
+  }));
+}
+
+export function getLockNameForCoursePath(coursePath) {
+  return `coursedir:${coursePath}`;
+}
+
+const CourseDataSchema = z.object({
+  path: z.string().nullable(),
+  branch: z.string().nullable(),
+  repository: z.string().nullable(),
+  commit_hash: z.string().nullable(),
+});
+
+export async function pullAndUpdate({
+  courseId,
+  userId,
+  authnUserId,
+  path,
+  branch,
+  repository,
+  commit_hash,
+}: {
+  courseId: string;
+  userId: string;
+  authnUserId: string;
+  path?: string | null;
+  branch?: string | null;
+  repository?: string | null;
+  commit_hash?: string | null;
+}): Promise<{ jobSequenceId: string; jobPromise: Promise<void> }> {
+  const serverJob = await createServerJob({
+    courseId,
+    userId,
+    authnUserId,
+    type: 'sync',
+    description: 'Pull from remote git repository',
+  });
+
+  const gitEnv = process.env;
+  if (config.gitSshCommand != null) {
+    gitEnv.GIT_SSH_COMMAND = config.gitSshCommand;
+  }
+
+  const jobPromise = serverJob.execute(async (job) => {
+    const lockName = getLockNameForCoursePath(path);
+    await namedLocks.tryWithLock(
+      lockName,
+      {
+        timeout: 5000,
+        onNotAcquired: () => {
+          job.fail('Another user is already syncing or modifying this course.');
+        },
+      },
+      async () => {
+        if (path === undefined || branch === undefined || repository === undefined) {
+          const course_data = await sqldb.queryRow(
+            sql.get_course_data,
+            { course_id: courseId },
+            CourseDataSchema,
+          );
```

Comment:
Can we get the whole course here using `CourseSchema` from `db-types`? We want to move towards using the standardized whole-row schemas (from `db-types`) as much as possible and not defining one-off schemas. Even better might be to add a `selectCourseById()` function that returns the whole-row object.

Reply:
Done in 2b78db3b4.